pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+
+
+Explicitly declaring implementation
+-----------------------------------
+
+To explicitly declare that a certain class implements the given protocols,
+they can be used as regular base classes. In this case a class could use
+default implementations of protocol members. ``typing.Sequence`` is a good
+example of a protocol with useful default methods.
+
+Abstract methods with trivial bodies are recognized by type checkers as
+having no default implementation and can't be used via ``super()`` in
+explicit subclasses. The default implementations can not be used if
+the subtype relationship is implicit and only via structural
+subtyping -- the semantics of inheritance is not changed. Examples::
+
+    class PColor(Protocol):
+        @abstractmethod
+        def draw(self) -> str:
+            ...
+        def complex_method(self) -> int:
+            # some complex code here
+
+    class NiceColor(PColor):
+        def draw(self) -> str:
+            return "deep blue"
+
+    class BadColor(PColor):
+        def draw(self) -> str:
+            return super().draw()  # Error, no default implementation
+
+    class ImplicitColor:   # Note no 'PColor' base here
+        def draw(self) -> str:
+            return "probably gray"
+        def comlex_method(self) -> int:
+            # class needs to implement this
+
+    nice: NiceColor
+    another: ImplicitColor
+
+    def represent(c: PColor) -> None:
+        print(c.draw(), c.complex_method())
+
+    represent(nice) # OK
+    represent(another) # Also OK
+
+Note that there is no conceptual difference between explicit and implicit
+subtypes, the main benefit of explicit subclassing is to get some protocol
+methods "for free". In addition, type checkers can statically verify that
+the class actually implements the protocol correctly::
+
+    class RGB(Protocol):
+        rgb: Tuple[int, int, int]
+
+        @abstractmethod
+        def intensity(self) -> int:
+            return 0
+
+    class Point(RGB):
+        def __init__(self, red: int, green: int, blue: str) -> None:
+            self.rgb = red, green, blue  # Error, 'blue' must be 'int'
+
+        # Type checker might warn that 'intensity' is not defined
+
+A class can explicitly inherit from multiple protocols and also form normal
+classes. In this case methods are resolved using normal MRO and a type checker
+verifies that all subtyping are correct. The semantics of ``@abstractmethod``
+is not changed, all of them must be implemented by an explicit subclass
+before it could be instantiated.
+
+
+Merging and extending protocols
+-------------------------------
+
+The general philosophy is that protocols are mostly like regular ABCs,
+but a static type checker will handle them specially. Subclassing a protocol
+class would not turn the subclass into a protocol unless it also has
+``typing.Protocol`` as an explicit base class. Without this base, the class
+is "downgraded" to a regular ABC that cannot be used with structural
+subtyping.
```

Comment:
I think we need a rationale for this rule. I think there probably is one, but it needs to be stated clearly, and I'm not sure what it is -- perhaps that we don't want to accidentally have some class act as a protocol just because one of its base classes happens to be one. That is, we still (slightly) prefer nominal subtyping over structural subtyping (in the static typing world).